### PR TITLE
Initial merge with xeus-clang-repl (mini)

### PR DESCRIPTION
**Pull request by alexander-penev**

### CODE REVIEW

File: src/xinterpreter.cpp [hunk 1]
```diff
@@ -243,13 +244,55 @@ void AddIncludePath(clang::Interpreter& Interp, llvm::StringRef Path)
...
+  std::string resource_dir = Cpp::DetectResourceDir();
+  ClangArgs.push_back("-resource-dir");
+  ClangArgs.push_back(resource_dir.c_str());
+  std::vector<std::string> CxxSystemIncludes;
+  Cpp::DetectSystemCompilerIncludePaths(CxxSystemIncludes);
+  for (const std::string& CxxInclude : CxxSystemIncludes) {
+    ClangArgs.push_back("-isystem");
+    ClangArgs.push_back(CxxInclude.c_str());
+  }
+  ClangArgs.push_back("-I/home/apenev/.conda/envs/xeus-cpp/include/");
```
Comment (vgvassilev) [inline src/xinterpreter.cpp:281]: We need to figure out how to get rid of this.

File: CMakeLists.txt [hunk 2]
```diff
@@ -155,10 +207,18 @@ set(XEUS_CPP_SRC
     src/xoptions.cpp
     src/xparser.cpp
     src/xutils.cpp
+    src/xdemangle.hpp
+    src/xinspect.hpp
+    src/xsystem.hpp
+    src/xparser.hpp
+    src/xmagics/os.hpp
+    src/xmagics/os.cpp
 )
```
Comment (vgvassilev) [inline CMakeLists.txt:216]: Let's make this a separate pull request.

File: CMakeLists.txt [hunk 3]
```diff
@@ -19,6 +19,12 @@ set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake;${CMAKE_MODULE_PATH}")
 
 set(XEUS_CPP_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/include)
 
+enable_language(CXX)
+set(CMAKE_CXX_EXTENSIONS NO)
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+
+include(GNUInstallDirs)
+
```
Comment (vgvassilev) [inline CMakeLists.txt:27]: This can go as a separate pull request.

File: CMakeLists.txt [hunk 4]
```diff
@@ -121,18 +114,77 @@ if(EMSCRIPTEN)
     set(EMSCRIPTEN_FEATURES "${EMSCRIPTEN_FEATURES} -s \"EXTRA_EXPORTED_RUNTIME_METHODS=[ENV']\"")
 endif()
 
-find_package(Clang REQUIRED)
-include(AddLLVM)
-include(HandleLLVMOptions)
-add_definitions(${LLVM_DEFINITIONS})
+find_package(CppInterOp REQUIRED CONFIG PATHS "${CMAKE_PREFIX_PATH}" "${CMAKE_PREFIX_PATH}/lib" "${CPPINTEROP_DIR}" "${CPPINTEROP_DIR}/lib")
```
Comment (vgvassilev) [inline CMakeLists.txt:117]: Do we need to have these extra paths? Maybe we can add them when we add the dockerfile.

File: CMakeLists.txt [hunk 5]
```diff
@@ -121,18 +114,77 @@ if(EMSCRIPTEN)
...
+if(CppInterOp_FOUND)
+  set(CppInterOp_CLANG_VERSION "17")
+  message(STATUS "Found CppInterOp: config=${CppInterOp_CONFIG} dir=${CppInterOp_DIR} (found version=${CppInterOp_VERSION} compatible with Clang ${CppInterOp_CLANG_VERSION}.x)")
```
Comment (vgvassilev) [inline CMakeLists.txt:120]: Maybe this message is not needed anymore?

File: CMakeLists.txt [hunk 6]
```diff
@@ -121,18 +114,77 @@ if(EMSCRIPTEN)
...
+endif()
+
 find_package(argparse REQUIRED)
 find_package(pugixml REQUIRED)
 
+set(Python_FIND_VIRTUALENV ONLY)
+find_package(Python COMPONENTS Interpreter Development)
```
Comment (vgvassilev) [inline CMakeLists.txt:127]: This should go.

File: CMakeLists.txt [hunk 7]
```diff
@@ -121,18 +114,77 @@ if(EMSCRIPTEN)
...
+      COPYONLY
+      )
+  endforeach()
+endfunction()
+
+file(GLOB _kernels LIST_DIRECTORIES true RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}" "share/jupyter/kernels/*")
+foreach(_kernel IN LISTS _kernels)
+  message("Configure kernel: ${_kernel}")
+  configure_kernel("${_kernel}")
+endforeach()
```
Comment (vgvassilev) [inline CMakeLists.txt:179]: This can go as a separate pull request. We should rename the introduced `CMAKE_` into `XEUS_CPP_` to avoid confusion.

File: CMakeLists.txt [hunk 8]
```diff
@@ -262,7 +322,11 @@ macro(xeus_cpp_create_target target_name linkage output_name)
         set(XEUS_CPP_XEUS_TARGET xeus-static)
     endif ()
 
-    target_link_libraries(${target_name} PUBLIC ${XEUS_CPP_XEUS_TARGET} clangInterpreter pugixml argparse::argparse xtl)
+    target_link_libraries(${target_name} PUBLIC ${XEUS_CPP_XEUS_TARGET} clangCppInterOp pugixml argparse::argparse xtl)
+    if (NOT EMSCRIPTEN)
+      target_link_libraries(${target_name} PUBLIC Python::Python)
+    endif()
+
```
Comment (vgvassilev) [inline CMakeLists.txt:329]:
```suggestion
```

File: CMakeLists.txt [hunk 9]
```diff
@@ -319,6 +383,30 @@ if (XEUS_CPP_BUILD_EXECUTABLE)
     xeus_cpp_set_common_options(xcpp)
     xeus_cpp_set_kernel_options(xcpp)
     target_link_libraries(xcpp PRIVATE xeus-zmq)
+    set_target_properties(xcpp PROPERTIES
+      ENABLE_EXPORTS 1
+      CXX_STANDARD ${CMAKE_CXX_STANDARD})
+    target_link_libraries(xcpp PUBLIC xeus-cpp pthread Python::Python)
```
Comment (vgvassilev) [inline CMakeLists.txt:389]: We probably do not need this.

File: CMakeLists.txt [hunk 10]
```diff
@@ -348,10 +436,10 @@ set(XEUS_CPP_CMAKECONFIG_INSTALL_DIR "${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NA
...
     install(TARGETS ${XEUS_CPP_TARGETS}
             EXPORT ${PROJECT_NAME}-targets
-            ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
-            LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-            RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
-            PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/xeus-cpp)
+            ARCHIVE DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}
+            LIBRARY DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}
+            RUNTIME DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_BINDIR}
+            PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_INCLUDEDIR}/xeus-cpp)
```
Comment (vgvassilev) [inline CMakeLists.txt:442]: Do we need these changes?

File: CMakeLists.txt [hunk 11]
```diff
@@ -319,6 +383,30 @@ if (XEUS_CPP_BUILD_EXECUTABLE)
...
+    target_link_libraries(xeus-cpp PUBLIC ${PYTHON_LIBRARIES})
+    target_link_libraries(xeus-cpp ${PYTHON_LIBRARIES_Development_Main})
+    set_target_properties(xeus-cpp PROPERTIES
+                      PUBLIC_HEADER "${XEUS_CPP_HEADERS}"
+                      COMPILE_DEFINITIONS "XEUS_CPP_EXPORTS"
+                      PREFIX ""
+                      VERSION ${${PROJECT_NAME}_VERSION}
+                      SOVERSION ${XEUS_CPP_VERSION_MAJOR}
+                      OUTPUT_NAME "libxeus-cpp"
+                      CXX_STANDARD ${CMAKE_CXX_STANDARD})
```
Comment (vgvassilev) [inline CMakeLists.txt:409]: That can probably go, too?

File: CMakeLists.txt (hunk 2 above)
Comment (vgvassilev) [reply CMakeLists.txt:211]: https://github.com/compiler-research/xeus-cpp/pull/48

File: CMakeLists.txt [hunk 12]
```diff
@@ -121,13 +111,67 @@ if(EMSCRIPTEN)
     set(EMSCRIPTEN_FEATURES "${EMSCRIPTEN_FEATURES} -s \"EXTRA_EXPORTED_RUNTIME_METHODS=[ENV']\"")
 endif()
 
-find_package(Clang REQUIRED)
-include(AddLLVM)
-include(HandleLLVMOptions)
-add_definitions(${LLVM_DEFINITIONS})
+find_package(CppInterOp REQUIRED CONFIG PATHS "${CMAKE_PREFIX_PATH}" "${CMAKE_PREFIX_PATH}/lib" "${CPPINTEROP_DIR}" "${CPPINTEROP_DIR}/lib")
+if(CppInterOp_FOUND)
+  set(CppInterOp_CLANG_VERSION "17")
```
Comment (vgvassilev) [inline CMakeLists.txt:116]: Shouldn't this go in the `CppInterOpConfig.cmake`?

File: test/pytest.ini [hunk 13]
```diff
@@ -1,3 +1,3 @@
 [pytest]
 testpaths = test
-addopts = --nbval --current-env
+addopts = --nbval --nbval-current-env
```
Comment (vgvassilev) [inline test/pytest.ini:3]: Do we need this change?

File: CMakeLists.txt [hunk 14]
```diff
@@ -362,16 +428,17 @@ endif ()
 
 # Install xcpp
 if (XEUS_CPP_BUILD_EXECUTABLE)
+message("!!! CMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX} CMAKE_INSTALL_BINDIR=${CMAKE_INSTALL_BINDIR}")
```
Comment (vgvassilev) [inline CMakeLists.txt:431]: Stray comment.

File: CMakeLists.txt [hunk 15]
```diff
@@ -362,16 +428,17 @@ endif ()
...
     install(TARGETS xcpp
-            RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
+            RUNTIME DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_BINDIR})
```
Comment (vgvassilev) [inline CMakeLists.txt:433]: Maybe we do not need this?

File: CMakeLists.txt [hunk 16]
```diff
@@ -121,13 +111,67 @@ if(EMSCRIPTEN)
...
+  set(XEUS_CPP_VENV_PATH "$ENV{VENV}")
+
+  message(STATUS "Debug: Replace in kernels")
+  message(STATUS "Debug: XEUS_CPP_CPLUS_INCLUDE_PATH=${XEUS_CPP_CPLUS_INCLUDE_PATH}")
+  message(STATUS "Debug: XEUS_CPP_PATH=${XEUS_CPP_PATH}")
+  message(STATUS "Debug: XEUS_CPP_LD_LIBRARY_PATH=${XEUS_CPP_LD_LIBRARY_PATH}")
+  message(STATUS "Debug: XEUS_CPP_PYTHONPATH=${XEUS_CPP_PYTHONPATH}")
+  message(STATUS "Debug: XEUS_CPP_VENV_PATH=${XEUS_CPP_VENV_PATH}")
+  message(STATUS "Debug: CMAKE_CURRENT_SOURCE_DIR=${CMAKE_CURRENT_SOURCE_DIR}")
+  message(STATUS "Debug: CMAKE_CURRENT_BINARY_DIR=${CMAKE_CURRENT_BINARY_DIR}")
```
Comment (vgvassilev) [inline CMakeLists.txt:145]: Maybe we should remove the debug information from this PR.

File: environment-dev.yml [hunk 17]
```diff
@@ -11,9 +11,9 @@ dependencies:
   - nlohmann_json=3.11.2
   - cppzmq
   - xtl
-  - clangdev >=16,<17
+  - CppInterOp
   - pugixml
-  - cpp-argparse
+  - cpp-argparse <3.0
```
Comment (mcbarton) [inline environment-dev.yml:16]: I wouldn't made this change. If you allow this to include version 3.0 then you can cover arm builds, as I updated the conda feedstock of cpp-argparse to include a Linux Arm build here https://github.com/conda-forge/cpp-argparse-feedstock/pull/5 , and am currently waiting for my PR in the xeus-cpp conda feedstock to be approved, so that xeus-cpp has arm recipes available here https://github.com/conda-forge/xeus-cpp-feedstock/pull/10

Comment (vgvassilev) [reply environment-dev.yml:16]: Can you do that change in a separate PR? I pinged the folks about the pending PR on xeus-cpp.

Comment (mcbarton) [reply environment-dev.yml:16]: Yes. I will wait for the xeus-cpp PR to be merged, and then put this change in as a separate PR.

Comment (vgvassilev) [reply environment-dev.yml:16]: Ok.

Comment (anutosh491) [reply environment-dev.yml:16]: Would be great if you could just reflect these versioning changes and the clangdev change in the `README.md` file too. It displays the following currently 
`mamba install notebook cmake cxx-compiler xeus-zmq nlohmann_json=3.11.2 cppzmq xtl jupyterlab clangdev=16 cpp-argparse pugixml doctest -c conda-forge`

File: include/xeus-cpp/xinterpreter.hpp [hunk 18]
```diff
@@ -16,7 +16,7 @@
 #include <string>
 #include <vector>
 
-#include <clang/Interpreter/Interpreter.h>
+#include "clang/Interpreter/CppInterOp.h" // from CppInterOp package
```
Comment (SylvainCorlay) [inline include/xeus-cpp/xinterpreter.hpp:19]: Deos CppInterop place a header in the `clang/Interpreter` include directory?

File: src/xinterpreter.cpp [hunk 19]
```diff
@@ -498,70 +322,19 @@ namespace xcpp
...
         p_cout_strbuf = std::cout.rdbuf();
         p_cerr_strbuf = std::cerr.rdbuf();
 
         std::cout.rdbuf(&m_cout_buffer);
         std::cerr.rdbuf(&m_cerr_buffer);
-
-        // Inject versions of printf and fprintf that output to std::cout
-        // and std::cerr (see implementation above).
-        inject_symbol("printf", llvm::pointerToJITTargetAddress(printf_jit), *m_interpreter);
-        inject_symbol("fprintf", llvm::pointerToJITTargetAddress(fprintf_jit), *m_interpreter);
```
Comment (anutosh491) [inline src/xinterpreter.cpp:555]: Just curious to know if/why we don't need these functions anymore ?

File: test/test_interpreter.cpp [hunk 20]
```diff
@@ -16,9 +16,26 @@
...
+        nl::json user_expressions = nl::json::object();
+        nl::json result = interpreter.execute_request(
+            code,
+            /*silent=*/false,
+            /*store_history=*/false,
+            user_expressions,
+            /*allow_stdin=*/false
+        );
+        REQUIRE(result["status"] == "ok");
+    }
```
Comment (anutosh491) [inline test/test_interpreter.cpp:33]: Added test looks good.

File: include/xeus-cpp/xinterpreter.hpp (hunk 18 above)
Comment (anutosh491) [reply include/xeus-cpp/xinterpreter.hpp:19]: Yeah I was thinking about this too !

**Comment by anutosh491** [comment]
I think this looks fine to start with (still need to go through the changes in CMakeLists.txt). Wrote a few comments. I was anyways unhappy with the special branching we had in `process_code` which was being used in `execute_request_impl` . That makes more sense now.

File: include/xeus-cpp/xinterpreter.hpp (hunk 18 above)
Comment (vgvassilev) [reply include/xeus-cpp/xinterpreter.hpp:19]: Nope, it is not but that's the design of the package - we should change that in the new versions.

File: src/xinterpreter.cpp (hunk 19 above)
Comment (vgvassilev) [reply src/xinterpreter.cpp:555]: We do have a better way to redirect i/o streams with CppInterOp. That will be part of a new PR as we discussed to do a minimal change with this one.

File: environment-dev.yml (hunk 17 above)
Comment (SylvainCorlay) [reply environment-dev.yml:16]: The feedstock PRs will be merged momentarily. I think that it is fine now to put `argparse <3.1` (so that we can take patch releases).

File: CMakeLists.txt (hunk 12 above)
Comment (vgvassilev) [reply CMakeLists.txt]: See https://github.com/compiler-research/CppInterOp/pull/214. Once we release CppInterOp we will use that instead of the `CppInterOp_CLANG_VERSION`.

File: environment-dev.yml (hunk 17 above)
Comment (SylvainCorlay) [reply environment-dev.yml:16]:
```suggestion
  - cpp-argparse <3.1
```

**Comment by SylvainCorlay** [comment]
> clang-tidy review says "All clean, LGTM! 👍"

Could we disable the clang tidy "LGTM" comments?

File: include/xeus-cpp/xinterpreter.hpp (hunk 18 above)
Comment (vgvassilev) [reply include/xeus-cpp/xinterpreter.hpp:19]: Ah, wait, are we having a common place for all headers? If everything is under include then it copies it there…

File: src/xinterpreter.cpp (hunk 19 above)
Comment (vgvassilev) [reply src/xinterpreter.cpp:555]: Where does std::clog writes to? We only redirect file descriptors 1 and 2.

**Comment by mcbarton** [comment]
@SylvainCorlay The following changes can be made to the README now that the feedstock has been updated (cannot seem to provide a link to the readme so are detailed below)
xeus-cpp has not been packaged for the mamba (or conda) package manager. 
Since it is now available to for multiple platforms through conda-forge this can be changed.

mamba install notebook cmake cxx-compiler xeus-zmq nlohmann_json=3.11.2 cppzmq xtl jupyterlab clangdev=16 cpp-argparse pugixml doctest -c conda-forge
Needs to be changed to include CppInterOp, and possibly the restriction to clangdev=16


The dependencies section table needs to be updated to include the dependency on CppinterOp, and updated restriction of cpp-argparse for main.

**Comment by SylvainCorlay** [comment]
> @SylvainCorlay The following changes can be made to the README now that the feedstock has been updated (cannot seem to provide a link to the readme so are detailed below)
> xeus-cpp has not been packaged for the mamba (or conda) package manager.
> Since it is now available to for multiple platforms through conda-forge this can be changed.

The mention that xeus-cpp has not been packaged should probably be changed.
The mention to cppinterop should be included in this PR I guess.

File: src/xinterpreter.cpp (hunk 19 above)
Comment (SylvainCorlay) [reply src/xinterpreter.cpp:555]: std::clog will print to stdout by default - so will be directed to the frontend.

Comment (vgvassilev) [reply src/xinterpreter.cpp:555]: Isn’t that what we should expect if a user type std::clog in a jupyter cell? If not we can make the redirection of std::clog go somewhere else.

File: CMakeLists.txt [hunk 21]
```diff
@@ -121,13 +111,54 @@ if(EMSCRIPTEN)
...
+      COPYONLY
+      )
+  endforeach()
+endfunction()
+
+file(GLOB _kernels LIST_DIRECTORIES true RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}" "share/jupyter/kernels/*")
+foreach(_kernel IN LISTS _kernels)
+  message("Configure kernel: ${_kernel}")
+  configure_kernel("${_kernel}")
+endforeach()
```
Comment (SylvainCorlay) [inline CMakeLists.txt:160]: Do we need to change the kernelspec generation logic as part of this PR?

Comment (vgvassilev) [reply CMakeLists.txt:160]: We can move it in a separate PR but that one seemed small enough already.

**Comment by vgvassilev** [comment]
> > clang-tidy review says "All clean, LGTM! 👍"
> 
> Could we disable the clang tidy "LGTM" comments?

We could but I find it useful for most PRs. That spares several clicks for the reviewers.

**Comment by vgvassilev** [comment]
@SylvainCorlay, can we make another review iteration?

File: include/xeus-cpp/xinterpreter.hpp (hunk 18 above)
Comment (SylvainCorlay) [reply include/xeus-cpp/xinterpreter.hpp:19]: I am still confused by why `CppInterOp.h`, which is a header from the CppInterOp package is under `PREFIX/include/clang`.

File: CMakeLists.txt (hunk 21 above)
Comment (SylvainCorlay) [reply CMakeLists.txt:160]: I think it would make sense since this is unrelated to the core of this PR.

File: include/xeus-cpp/xinterpreter.hpp (hunk 18 above)
Comment (vgvassilev) [reply include/xeus-cpp/xinterpreter.hpp:19]: Yes, not a good choice. We will fix this in the next release of CppInterOp. Do we want to block progress because of that?

File: CMakeLists.txt (hunk 21 above)
Comment (vgvassilev) [reply CMakeLists.txt]: Done!

**Review by JohanMabille — APPROVED** [review]

**Review by anutosh491 — APPROVED** [review]

**Comment by vgvassilev** [comment]
@SylvainCorlay, can we move forward here? Is there something else we should improve as part of this pull request?

**Review by SylvainCorlay — APPROVED** [review]

**Comment by SylvainCorlay** [comment]
Merged!

File: share/jupyter/kernels/xcpp/kernel.json.in [hunk 22]
```diff
@@ -1,9 +1,16 @@
 {
   "display_name": "cpp 14 (xcpp)",
+  "env": {
+      "PATH":"@XEUS_CPP_PATH@",
+      "LD_LIBRARY_PATH":"@XEUS_CPP_LD_LIBRARY_PATH@"
```
Comment (SylvainCorlay) [inline share/jupyter/kernels/xcpp/kernel.json.in:5]: I merged, but I wonder why we need to set these up exactly, if it wasn't required before?

File: include/xeus-cpp/xinterpreter.hpp (hunk 18 above)
Comment (SylvainCorlay) [reply include/xeus-cpp/xinterpreter.hpp:19]: > We will fix this in the next release of CppInterOp.

Thanks!

File: CMakeLists.txt [hunk 23]
```diff
@@ -121,13 +111,29 @@ if(EMSCRIPTEN)
...
+    set(XEUS_CPP_KERNELSPEC_PATH "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_BINDIR}/")
+endif ()
+
+set(XEUS_CPP_PATH "$ENV{PATH}")
+set(XEUS_CPP_LD_LIBRARY_PATH "$ENV{LD_LIBRARY_PATH}")
+set(XEUS_CPP_RESOURCE_DIR ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}/clang/${CppInterOp_CLANG_VERSION})
+configure_file (
+    "${CMAKE_CURRENT_SOURCE_DIR}/share/jupyter/kernels/xcpp/kernel.json.in"
+    "${CMAKE_CURRENT_BINARY_DIR}/share/jupyter/kernels/xcpp/kernel.json"
+)
```
Comment (SylvainCorlay) [inline CMakeLists.txt:135]: Wha is the benefit of this version compared to the previous one?

(trying to understand better because keeping cmake files of xeus-based kernels similar makes it easier to make changes across all kernels).
